pkg/api/v1/state.go

```diff
@@ -0,0 +1,232 @@
+	for _, dashboard := range cfg.Dashboards {
+		if config.Normalize(dashboard.Name) == dashboardKey {
+			for _, tab := range dashboard.DashboardTab {
+				if config.Normalize(tab.Name) == tabKey {
+					return tab
+				}
+			}
+			return nil
+		}
+	}
+	return nil
+}
+
+// getTabGrid fetch tab group name grid info (columns, rows, ..etc)
+func (s Server) getTabGrid(configPath *gcs.Path, cfg *configpb.Configuration, groupName string) (*statepb.Grid, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Minute) // todo(abdelrady): opt.timeout setup instead of fixed 10
+	defer cancel()
+	groupFinder := func(name string) (*configpb.TestGroup, gridReader, error) {
+		group := config.FindTestGroup(name, cfg)
+		if group == nil {
+			return nil, nil, nil
+		}
+		groupPath, err := configPath.ResolveReference(&url.URL{Path: path.Join(s.GridPathPrefix, name)})
+		if err != nil {
+			return group, nil, err
+		}
+		reader := func(ctx context.Context) (io.ReadCloser, time.Time, int64, error) {
+			return pathReader(ctx, s.Client, *groupPath)
+		}
+		return group, reader, nil
+	}
+
+	group, groupReader, err := groupFinder(groupName)
+	if err != nil || group == nil {
+		return nil, errors.New(fmt.Sprintf("not found: %q", groupName))
+	}
+	grid, _, _, err := readGrid(ctx, groupReader)
+	if err != nil {
+		return nil, errors.New(fmt.Sprintf("load %s: %v", groupName, err))
+	}
+
+	return grid, err
+}
+
+func (s Server) getDashboardTabGrid(r *http.Request, cfg *configpb.Configuration, dashboardKey string, tabKey string) (*statepb.Grid, string) {
+	tab := getDashboardTabInternal(cfg, dashboardKey, tabKey)
+	if tab == nil {
+		return nil, fmt.Sprintf("Dashboard {%q} or tab {%q} not found", dashboardKey, tabKey)
+	}
+	configPath, _, _ := s.configPath(r)
+	grid, err := s.getTabGrid(configPath, cfg, tab.TestGroupName)
+	if err != nil {
+		return nil, fmt.Sprintf("%v", err)
+	}
+	return grid, ""
+}
+
+// decodeRLE decode the run length encoded data
+//   [0, 3, 5, 4] -> [0, 0, 0, 5, 5, 5, 5]
+func decodeRLE(encodedData []int32) []int32 {
+	var decodedResult []int32
+	encodedDataLength := len(encodedData)
+	if encodedDataLength%2 == 0 {
+		for encodedDataIdx := 0; encodedDataIdx < encodedDataLength; encodedDataIdx += 2 {
+			for cellRepeatCount := encodedData[encodedDataIdx+1]; cellRepeatCount > 0; cellRepeatCount-- {
+				decodedResult = append(decodedResult, encodedData[encodedDataIdx])
+			}
+		}
+	}
+	return decodedResult
+}
+
+// GetTabHeaders returns dashboard tab headers
+// Response json: ListHeadersResponse
+func (s Server) GetTabHeaders(w http.ResponseWriter, r *http.Request) {
+	vars := mux.Vars(r)
+	cfg := s.getConfig(w, r)
+	if cfg == nil {
+		return
+	}
+
+	grid, errMsg := s.getDashboardTabGrid(r, cfg, vars["dashboard"], vars["tab"])
+	if errMsg != "" {
+		http.Error(w, errMsg, http.StatusNotFound)
+	}
+	if grid != nil && grid.Columns != nil {
```

Comment:
Prefer early error returns (https://danp.net/posts/reducing-go-nesting/); check if nil and return if so.

Also, not sure if grid.Columns = nil is an error; think we can just return an empty response instead if the columns or grid are empty, instead of not found.

Reply:
That makes sense. Updated.